processing/src/main/java/org/apache/druid/query/planning/DataSourceAnalysis.java

```diff
@@ -125,6 +126,11 @@ public static DataSourceAnalysis forDataSource(final DataSource dataSource)
       current = subQuery.getDataSource();
     }
 
+    while (current instanceof UnnestDataSource) {
```

Comment:
If there is a Query of an Unnest of a Query of an Unnest, the way that you have interleaved these is not going to completely unwrap the objects as expected.

This DataSourceAnalysis thing is probably another thing to move onto the DataSource object itself...  Not sure if we should do that now or leave it as something to do for later though.  either way, you need both conditions (check for Query and check for Unnest) on the while loop above.